app/vmselect/netstorage/tenant_filters.go

```diff
@@ -0,0 +1,141 @@
+package netstorage
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/searchutils"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/auth"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/prompbmarshal"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promrelabel"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/querytracer"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/storage"
+)
+
+// GetTenantTokensFromFilters returns the list of tenant tokens and the list of filters without tenant filters.
+func GetTenantTokensFromFilters(qt *querytracer.Tracer, tr storage.TimeRange, tfs [][]storage.TagFilter, deadline searchutils.Deadline) ([]storage.TenantToken, [][]storage.TagFilter, error) {
+	tenants, err := TenantsCached(qt, tr, deadline)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot obtain tenants: %w", err)
+	}
+
+	tenantFilters, otherFilters := splitFiltersByType(tfs)
+
+	tts, err := applyFiltersToTenantsStr(tenants, tenantFilters)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot apply filters to tenants: %w", err)
+	}
+
+	return tts, otherFilters, nil
+}
+
+func splitFiltersByType(tfs [][]storage.TagFilter) ([]string, [][]storage.TagFilter) {
```

Comment:
I think, it's better to keep filters signature as is:

`func splitFiltersByType(tfs [][]storage.TagFilter) ([]storage.TagFilter, [][]storage.TagFilter)`

And convert TagFilter at `applyFiltersToTenants`